node/subsystem-util/src/inclusion_emulator/staging.rs

```diff
@@ -0,0 +1,1226 @@
+//!
+//! Avoiding fragment-tree blowup is beyond the scope of this module.
+//!
+//! ### Pruning Fragment Trees
+//!
+//! When the relay-chain advances, we want to compare the new constraints
+//!    of that relay-parent to the roots of the fragment trees we have. There are 3 cases.
+//!
+//! 1. The root fragment is still valid under the new constraints. In this case, we do nothing.
+//!    This is the "prediction still uncertain" case.
+//! 2. The root fragment is invalid under the new constraints because it has been subsumed by the relay-chain.
+//!    in this case, we can discard the root and split & re-root the fragment tree
+//!    under its descendents and compare to the new constraints again.
+//!    This is the "prediction came true" case.
+//! 3. The root fragment is invalid under the new constraints because a competing parachain block has been included
+//!    or it would never be accepted for some other reason. In this case we can discard the entire
+//!    fragment tree.
+//!    This is the "prediction came false" case.
+//!
+//! This is all a bit of a simplification because it assumes that the relay-chain advances without
+//! forks and is finalized instantly. In practice, the set of fragment-trees needs to be observable
+//! from the perspective of a few different possible forks of the relay-chain and not pruned
+//! too eagerly.
+//!
+//! Note that the fragments themselves don't need to change and the only thing we care about
+//! is whether the predictions they represent are still valid.
+//!
+//! ### Extending Fragment Trees
+//!
+//! As predictions fade into the past, new ones should be stacked on top.
+//!
+//! Every new relay-chain block is an opportunity to make a new prediction about the future.
+//! higher-level logic should select the leaves of the fragment-trees to build upon or whether
+//! to create a new fragment-tree.
+//!
+//! ### Code Upgrades
+//!
+//! Code upgrades are the main place where this emulation fails. The on-chain PVF upgrade scheduling
+//! logic is very path-dependent and intricate so we just assume that code upgrades
+//! can't be initiated and applied within a single fragment-tree. Fragment-trees aren't deep,
+//! in practice and code upgrades are fairly rare. So what's likely to happen around code
+//! upgrades is that the entire fragment-tree has to get discarded at some point.
+//!
+//! That means a few blocks of execution time lost, which is not a big deal for code upgrades
+//! in practice at most once every few weeks.
+
+use polkadot_primitives::vstaging::{
+	BlockNumber, CandidateCommitments, CollatorId, CollatorSignature, Hash, HeadData, Id as ParaId,
+	PersistedValidationData, UpgradeRestriction, ValidationCodeHash,
+};
+use std::collections::HashMap;
+
+/// Constraints on inbound HRMP channels.
+#[derive(Debug, Clone, PartialEq)]
+pub struct InboundHrmpLimitations {
+	/// An exhaustive set of all valid watermarks, sorted ascending
+	pub valid_watermarks: Vec<BlockNumber>,
```

Comment:
We should really strive for making sorted assumptions more explict. Would a `BTreeSet` work here?